### PR TITLE
Add support for `pre-wrap` and `pre-line` values for `white-space`.

### LINKED ISSUE #1513: implement white-space
Needed for #1366.

### PR DESCRIPTION
**Pull request by eefriedman**
This is mostly straightforward.  I had to modify a couple of places
which were accidentally discarding whitespace.

Fixes #1513.

This fixes some relevant tests from the CSS testsuite... but a lot of
them are either manual, or don't pass because of unrelated issues.  (For
example, white-space-mixed-002 renders correctly, but
white-space-mixed-002-ref doesn't because of a float bug.)

I'd appreciate any suggestions for how to go about adding tests for this.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7951)

### CODE REVIEW

File: components/gfx/text/util.rs [hunk 1]
```diff
@@ -38,7 +38,7 @@ pub fn transform_text(text: &str,
                     output_text.push(ch);
                 }
             }
-            text.len() > 0 && is_in_whitespace(text.char_at_reverse(0), mode)
+            false
```
Comment (pcwalton) [inline components/gfx/text/util.rs]: Why this change?

File: components/layout/fragment.rs [hunk 2]
```diff
@@ -1275,10 +1275,18 @@ impl Fragment {
...
+        if !self.is_scanned_text_fragment() {
+            return false;
+        }
+        match self.style.get_inheritedtext().white_space {
+            white_space::T::pre |
+            white_space::T::nowrap => false,
+            white_space::T::normal |
+            white_space::T::pre_wrap |
+            white_space::T::pre_line => true,
+        }
```
Comment (pcwalton) [inline components/layout/fragment.rs]: Matches very similar to this are duplicated throughout the code. Maybe we should have a `whitespace_is_pre_or_nowrap()` kind of function?

Comment (pcwalton) [inline components/layout/fragment.rs]: Actually, never mind, I see the matches are subtly different each time.

File: components/layout/fragment.rs [hunk 3]
```diff
@@ -1560,18 +1570,9 @@ impl Fragment {
...
-                    inline_start_range.shift_by(slice.range.length());
-                } else {
-                    debug!("calculate_split_position_using_breaking_strategy: enlarging span");
-                    remaining_inline_size = remaining_inline_size - advance;
-                    inline_start_range.extend_by(slice.range.length());
-                }
-                pieces_processed_count += 1;
+                debug!("calculate_split_position_using_breaking_strategy: enlarging span");
+                remaining_inline_size = remaining_inline_size - advance;
+                inline_start_range.extend_by(slice.range.length());
```
Comment (pcwalton) [inline components/layout/fragment.rs]: Why unconditionally set `remaining_inline_size` now?

**Comment by pcwalton** [comment]
Looks good other than these couple of questions I had.

File: components/gfx/text/util.rs (hunk 1 above)
Comment (eefriedman) [inline components/gfx/text/util.rs]: This fixes https://github.com/servo/servo/blob/master/tests/wpt/css-tests/css21_dev/html4/white-space-mixed-002.htm and some related issues.  The return value here basically means "should leading whitespace be discarded from the next fragment", and the answer is always no in `white-space: pre` or `white-space: pre-wrap` mode.

File: components/layout/fragment.rs (hunk 2 above)
Comment (eefriedman) [inline components/layout/fragment.rs]: Hmm... there are basically three useful ways to divide up the values of white-space: "are spaces preserved", "are newlines preserved", and "is wrapping allowed".  I could add helpers for all three, which might be useful for readability even if they don't have very many uses.

File: components/layout/fragment.rs (hunk 3 above)
Comment (eefriedman) [inline components/layout/fragment.rs]: Like the debug! log said, the first branch would trim leading whitespace.  That doesn't really work out very well in pre-wrap mode, where whitespace is supposed to be preserved.

I was going to add the condition that the fragment isn't using a whitespace-preserving mode... but then I realized we never actually end up in that situation: we strip leading whitespace before we reach this code.

File: components/layout/fragment.rs (hunk 2 above)
Comment (pcwalton) [inline components/layout/fragment.rs]: Yeah, I agree those would be useful distinctions for readability.

**Comment by Ms2ger** [comment]
This needs at least some automated testing. If you know which csswg-test tests this causes to pass, feel free to submit references upstream and ask me to review. We can land this PR once I've integrated those references.

**Comment by eefriedman** [comment]
Refactored the code a bit, and added some tests.

**Comment by pcwalton** [comment]
Looks great, thanks!

**Comment by eefriedman** [comment]
Looks like a legitimate bug in my patch.

**Comment by eefriedman** [comment]
Added a fix for newlines in `white-space: pre` mode.  (See https://github.com/servo/servo/pull/7951/files#diff-013814e2f277c3fe9e1b72e33bd8ad2aL621)  Patch should be good to go now.

**Comment by eefriedman** [comment]
@pcwalton Ping.  This is ready to merge.

**Comment by eefriedman** [comment]
Rebased.

**Comment by eefriedman** [comment]
Rebased again.

@pcwalton ping.

**Comment by pcwalton** [comment]
Sorry this fell through the cracks!
